cloudinit/distros/debian.py

```diff
@@ -277,31 +286,132 @@ def read_system_locale(sys_path=LOCALE_CONF_FN, keyname="LANG"):
     return sys_val
 
 
+def _ensure_tool(bin_name: str, install_function=None, pkgs=None):
+    if subp.which(bin_name):
+        return
+    if install_function and pkgs:
+        install_function(pkgs)
+    # if not present and installed raise no RuntimeError
+    # but wait for subp to fail
+
+
+def _normalize_locale(requested: str) -> str:
+    if not requested:
+        raise ValueError("Failed to provide locale value.")
+    req = requested.strip()
+
+    # Accept canonical “no-regeneration” values as-is
+    if req.lower() in ("c", "posix", "c.utf-8", "c.utf8"):
+        return (
+            "C.UTF-8" if req.lower() != "c" and "utf" in req.lower() else "C"
+        )
+
+    # If no charset specified, default to UTF-8
+    if "." not in req and "@" not in req:
+        return req + ".UTF-8"
+    return req
+
+
 def update_locale_conf(
-    locale, sys_path, keyname="LANG", install_function=None
+    locale, sys_path, default_locale, keyname="LANG", install_function=None
 ):
     """Update system locale config"""
     LOG.debug(
         "Updating %s with locale setting %s=%s", sys_path, keyname, locale
     )
-    if not subp.which("update-locale"):
-        install_function(["locales"])
+    _ensure_tool("update-locale", install_function, ["locales"])
     subp.subp(
         [
             "update-locale",
             "--locale-file=" + sys_path,
             "%s=%s" % (keyname, locale),
         ],
+        update_env={
+            "LANGUAGE": default_locale,
+            "LANG": default_locale,
+            "LC_ALL": default_locale,
+        },
         capture=False,
     )
 
 
-def regenerate_locale(locale, sys_path, keyname="LANG", install_function=None):
+def _lookup_supported_line(requested: str) -> str:
+    """
+    Return the canonical line from /usr/share/i18n/SUPPORTED for `requested`.
+
+    Accepts:
+      - bare language_region:   "fi_FI"
+      - with charset:           "fi_FI.ISO-8859-1" or "fi_FI.UTF-8"
+      - with modifier:          "fi_FI@euro" (works with/without charset)
+
+    Prefers UTF-8 only when the request didn’t specify a charset and multiple
+    candidates exist; otherwise returns the first match.
     """
-    Run locale-gen for the provided locale and set the default
-    system variable `keyname` appropriately in the provided `sys_path`.
+    try:
+        sup = util.load_text_file(SUPPORTED_FN).splitlines()
+    except Exception:
```

Comment:
Please avoid excepts which use the Exception type. We want to handle specific errors or fail hard with unexpected other types of exceptions.

Here I think we can use FileNotFoundError
```suggestion
    except FileNotFoundError:
```